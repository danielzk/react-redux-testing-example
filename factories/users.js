import {Factory} from 'rosie'
import {OrderedMap} from 'immutable'

export const UserFactory = new Factory()
  .sequence('id', sequence => sequence)
  .sequence('username', sequence => 'username' + sequence)
  .attr('firstName', 'Mark')
  .attr('lastName', 'Zucaritas')

export const UserListFactory = (count) => {
  const users = UserFactory.buildList(count)
  return OrderedMap(users.map(item => [item.id, item]))
}
