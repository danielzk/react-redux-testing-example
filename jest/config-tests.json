{
  "setupTestFrameworkScriptFile": "./jest/setup",
  "verbose": true,
  "rootDir": "../",
  "snapshotSerializers": [
    "enzyme-to-json/serializer"
  ],
  "collectCoverage": true,
  "collectCoverageFrom": [
    "**/*.{js}",
    "!**/node_modules/**",
    "!**webpack*.js",
    "!.eslintrc.js"
  ],
  "coverageDirectory": "./jest/.coverage"
}
