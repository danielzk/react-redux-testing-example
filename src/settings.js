export const urls = {
  userDetail: (id) => `/api/v1/users/${id}/`,
  userList: () => 'api/v1/users/'
}
