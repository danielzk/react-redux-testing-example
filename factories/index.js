import {Factory} from 'rosie'

import {UserFactory, UserListFactory} from './users'

export const StoreFactory = new Factory()
  .option('numUsers', 2)
  .attr('users', ['numUsers'], count => UserListFactory(count))

export {UserFactory, UserListFactory}
