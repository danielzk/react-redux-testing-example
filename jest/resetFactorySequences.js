import * as factories from '../factories'

export default () => {
  Object.values(factories).forEach((Factory) => Factory.sequences = {})
}
