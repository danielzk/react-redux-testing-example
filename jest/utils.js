import * as factories from '../factories'

export const resetFactorySequences = () => {
  Object.values(factories).forEach((Factory) => Factory.sequences = {})
}

export const getSnapshotRequest = (response) => {
  return {
    method: response.config.method,
    url: response.config.url,
    data: response.config.data ? JSON.parse(response.config.data) : null,
  }
}
