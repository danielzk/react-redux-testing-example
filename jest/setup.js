import 'babel-polyfill'
import Adapter from 'enzyme-adapter-react-16'
import {configure} from 'enzyme'

import {resetFactorySequences} from './utils'

beforeEach(() => resetFactorySequences())

configure({adapter: new Adapter()})
