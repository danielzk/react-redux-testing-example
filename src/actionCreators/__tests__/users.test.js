import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import {createUser, deleteUser, loadUsers} from '../'
import {urls} from '../../settings'
import {StoreFactory, UserFactory} from '../../../factories'
import {getSnapshotRequest} from '../../../jest/utils'

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

describe('Users actionCreators', () => {
  let axiosMock = new MockAdapter(axios)

  beforeEach(() => {
    axiosMock.reset()
  })

  it('createUser should create the user and create ADD_USER action', async () => {
    const user = UserFactory.build()
    axiosMock.onPost(urls.userList()).reply(201, user)
    const store = mockStore(StoreFactory.build())
    store.dispatch(createUser(user.username, user.firstName, user.lastName))
      .then((response) => {
        expect(store.getActions()).toMatchSnapshot('dispatched actions')
        expect(getSnapshotRequest(response)).toMatchSnapshot('request data')
      })
  })

  it('deleteUser should delete the user and create REMOVE_USER action', async () => {
    axiosMock.onDelete(urls.userDetail(1)).reply(204)
    const store = mockStore(StoreFactory.build())
    store.dispatch(deleteUser(1))
      .then((response) => {
        expect(store.getActions()).toMatchSnapshot('dispatched actions')
        expect(getSnapshotRequest(response)).toMatchSnapshot('request data')
      })
  })

  it('loadUsers should load users and create SET_USERS action', async () => {
    axiosMock.onGet(urls.userList()).reply(200, UserFactory.buildList(2))
    const store = mockStore(StoreFactory.build())
    store.dispatch(loadUsers())
      .then((response) => {
        expect(store.getActions()).toMatchSnapshot('dispatched actions')
        expect(getSnapshotRequest(response)).toMatchSnapshot('request data')
      })
  })
})
