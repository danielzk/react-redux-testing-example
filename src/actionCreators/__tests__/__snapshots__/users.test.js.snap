// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users actionCreators createUser should create the user and create ADD_USER action: dispatched actions 1`] = `
Array [
  Object {
    "payload": Object {
      "user": Object {
        "firstName": "Mark",
        "id": 1,
        "lastName": "Zucaritas",
        "username": "username1",
      },
    },
    "type": "ADD_USER",
  },
]
`;

exports[`Users actionCreators createUser should create the user and create ADD_USER action: request data 1`] = `
Object {
  "data": Object {
    "firstName": "Mark",
    "lastName": "Zucaritas",
    "username": "username1",
  },
  "method": "post",
  "url": "/api/v1/users/",
}
`;

exports[`Users actionCreators deleteUser should delete the user and create REMOVE_USER action: dispatched actions 1`] = `
Array [
  Object {
    "payload": Object {
      "id": 1,
    },
    "type": "REMOVE_USER",
  },
]
`;

exports[`Users actionCreators deleteUser should delete the user and create REMOVE_USER action: request data 1`] = `
Object {
  "data": null,
  "method": "delete",
  "url": "/api/v1/users/1/",
}
`;

exports[`Users actionCreators loadUsers should load users and create SET_USERS action: dispatched actions 1`] = `
Array [
  Object {
    "payload": Object {
      "users": Immutable.OrderedMap {
        1: Object {
          "firstName": "Mark",
          "id": 1,
          "lastName": "Zucaritas",
          "username": "username1",
        },
        2: Object {
          "firstName": "Mark",
          "id": 2,
          "lastName": "Zucaritas",
          "username": "username2",
        },
      },
    },
    "type": "SET_USERS",
  },
]
`;

exports[`Users actionCreators loadUsers should load users and create SET_USERS action: request data 1`] = `
Object {
  "data": null,
  "method": "get",
  "url": "/api/v1/users/",
}
`;
